src/app.ts

```diff
@@ -9,20 +9,46 @@ import type { FastifyRequest, FastifyReply } from 'fastify'
 
 const w = usb()
 
+const validSubprotocols: string[] = ['XAPWS1']
+
 const app = fastify({ logger: true })
-app.register(fastifyws)
-
-app.get('/ws', { websocket: true }, (connection: SocketStream, req: FastifyRequest) => {
-  connection.socket.on('message', (message: MessageEvent) => {
-    w.postMessage(message.toString())
-    w.once('message', (result) => {
-      req.log.info(result)
-      connection.socket.send(result)
-    })
-    w.on('error', (err) => {
-      req.log.error(err)
-      connection.socket.send(err.toString())
-      process.exit(1)
+app.register(fastifyws, {
+  options: {
+    /**
+     * Specify how to format sec-websocket-protocol header on upgrade
+     * @param protocols set of subprotocols from header
+     * @returns subprotocol accepted or false
+     */
+    handleProtocols: (protocols: Set<string>) => {
```

Comment:
i'd replace `handleProtocols` with:
```ts
    verifyClient: (info: { origin: string; req: IncomingMessage; secure: boolean }) => {
      return (
        (info.req.headers['sec-websocket-protocol'] &&
          validSubprotocols.find((x) => info.req.headers['sec-websocket-protocol'] === x)) ||
        false
      )
    }
```
If there is no subprotocol this will not be executed. See: https://github.com/websockets/ws/issues/1552#issuecomment-483642437

Reply:
In the notes they recommended not using this hook. I wasn't sure why

Reply:
Handle protocols only sets the headers as per the spec. It doesn't seem to care about verification

Reply:
we can have prevalidation, so we need to use `app.route` and add `prehandler` on it. We cannot use the shorthand route definition like `app.get`